src/codec.rs

```diff
@@ -488,29 +550,99 @@ impl<T, X> Encode for X where
 pub trait WrapperTypeDecode: Sized {
 	/// A wrapped type.
 	type Wrapped: Into<Self>;
+
+	// !INTERNAL USE ONLY!
+	// This is a used to specialize `decode` for the wrapped type.
+	#[doc(hidden)]
+	#[inline]
+	fn decode_wrapped<I: Input>(input: &mut I) -> Result<Self, Error> where Self::Wrapped: Decode {
+		input.descend_ref()?;
+		let result = Ok(Self::Wrapped::decode(input)?.into());
+		input.ascend_ref();
+		result
+	}
 }
+
 impl<T> WrapperTypeDecode for Box<T> {
 	type Wrapped = T;
+
+	fn decode_wrapped<I: Input>(input: &mut I) -> Result<Self, Error> where Self::Wrapped: Decode {
+		input.descend_ref()?;
+
+		// Placement new is not yet stable, but we can just manually allocate a chunk of memory
+		// and convert it to a `Box` ourselves.
+		//
+		// The explicit types here are written out for clarity.
+		//
+		// TODO: Use `Box::new_uninit` once that's stable.
+		let layout = core::alloc::Layout::new::<MaybeUninit<T>>();
+
+		let ptr: *mut MaybeUninit<T> = if layout.size() == 0 {
+			core::ptr::NonNull::dangling().as_ptr()
+		} else {
+
+			// SAFETY: Layout has a non-zero size so calling this is safe.
+			let ptr: *mut u8 = unsafe {
+				crate::alloc::alloc::alloc(layout)
+			};
+
+			if ptr.is_null() {
+				crate::alloc::alloc::handle_alloc_error(layout);
+			}
+
+			ptr.cast()
+		};
+
+		// SAFETY: Constructing a `Box` from a piece of memory allocated with `std::alloc::alloc`
+		//         is explicitly allowed as long as it was allocated with the global allocator
+		//         and the memory layout matches.
+		//
+		//         Constructing a `Box` from `NonNull::dangling` is also always safe as long
+		//         as the underlying type is zero-sized.
+		let mut boxed: Box<MaybeUninit<T>> = unsafe { Box::from_raw(ptr) };
+
+		T::decode_into(DecodeContext::new(), input, &mut boxed)?;
+
+		// Decoding succeeded, so let's get rid of `MaybeUninit`.
+		let ptr: *mut MaybeUninit<T> = Box::into_raw(boxed);
+		let ptr: *mut T = ptr.cast();
+
+		// SAFETY: `MaybeUninit` doesn't affect the memory layout, so casting the pointer back
+		//         into a `Box` is safe.
+		let boxed: Box<T> = unsafe { Box::from_raw(ptr) };
+
+		input.ascend_ref();
+		Ok(boxed)
+	}
 }
+
 impl<T> WrapperTypeDecode for Rc<T> {
 	type Wrapped = T;
+
+	fn decode_wrapped<I: Input>(input: &mut I) -> Result<Self, Error> where Self::Wrapped: Decode {
+		// TODO: This is inefficient; use `Rc::new_uninit` once that's stable.
+		Box::<T>::decode(input).map(|output| output.into())
```

Comment:
> Could we not just put the code for the `Box` implementation into some macro and reuse it here? see_no_evil

That'd be tricky because without `::new_uninit` we have no good way to preallocate the uninitialized memory for the `Rc`/`Arc`.

In case of a `Box<T>` the only thing that the memory holds is a `T`, so we can get its layout and manually allocate the memory.

In case of `Rc`/`Arc` there are also the refcounts allocated in the same chunk of memory, and there's AFAIK there's no good way to get a `Layout` that could be used to manually preallocate that memory and pass it to `Rc::from_raw`. One way we could do it would be to just hardcode it (this is basically equivalent `Box<(usize, usize, T)>`), and then verify in a build script that it's correct (in case it suddenly changes, since it's an internal implementation detail), and if it is then do it the fast way, and if it isn't then emit a compile-time warning and do it the slow way. So for all practical purposes that'd work and be safe, but obviously it is very hacky and I don't really like it. And I'm not convinced it's worth it. (As opposed to just waiting waiting until the `new_uninit` methods are stabilized.)